### PR TITLE
Add Lichess OAuth

### PR DESCRIPTION
**Pull request by Waxmatuct**
The authorization code flow with PKCE that allows users to login with Lichess. [https://lichess.org/api#tag/OAuth](url)

### CODE REVIEW

File: src/Lichess/Provider.php [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+     * @param string $token
+     *
+     * @return string
+     */
+    protected function getEmailByToken($token)
+    {
+        $url = 'https://lichess.org/api/account/email';
+
+        $response = $this->getHttpClient()->get($url, [
+            'headers' => [
```
Comment (lucasmichot) [inline src/Lichess/Provider.php:101]: Why not re-using `getRequestOptions` here ?

File: src/Lichess/Provider.php [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                'Authorization' => 'Bearer '.$token,
+            ],
+        ]);
+
+        $email = json_decode($response->getBody(), true);
+
+        return Arr::get($email, 'email');
```
Comment (lucasmichot) [inline src/Lichess/Provider.php:108]: `return json_decode((string) $response->getBody(), true)['email'] ?? null;` would be just fine

File: src/Lichess/Provider.php [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+        $response = $this->getHttpClient()->get(
+            $userUrl,
+            $this->getRequestOptions($token)
+        );
+
+        $user = json_decode($response->getBody(), true);
+
+        if (in_array('email:read', $this->scopes)) {
+            $user = Arr::prepend($user, $this->getEmailByToken($token), 'email');
```
Comment (lucasmichot) [inline src/Lichess/Provider.php:67]:
```suggestion
            $user['email'] = $this->getEmailByToken($token);
```

File: src/Lichess/Provider.php [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    protected function getUserByToken($token)
+    {
+        $userUrl = 'https://lichess.org/api/account';
+
+        $response = $this->getHttpClient()->get(
+            $userUrl,
+            $this->getRequestOptions($token)
+        );
+
+        $user = json_decode($response->getBody(), true);
```
Comment (lucasmichot) [inline src/Lichess/Provider.php:64]:
```suggestion
        $user = json_decode((string) $response->getBody(), true);
```

**Comment by Waxmatuct** [comment]
I don't know why, but now I can't obtain auth. I get Laravel\Socialite\Two\InvalidStateException Touch nothing. Laravel 8.53.1, PHP 7.4.18

**Review by atymic — APPROVED** [review]

**Comment by atymic** [comment]
https://packagist.org/packages/socialiteproviders/lichess

**Pull request merged**